react/features/vad-reporter/TrackVADEmitter.js

```diff
@@ -0,0 +1,281 @@
+        let closureRnnoiseProc = null;
+        let closureLocalTrack = null;
+
+        logger.log(`Initializing TrackVADEmitter for device: ${micDeviceId}`);
+
+        return createRnnoiseProcessor()
+            .then(rnnoiseProcessor => {
+                closureRnnoiseProc = rnnoiseProcessor;
+            })
+            .then(() =>
+                TrackVADEmitter._getAudioStream(micDeviceId)
+                    .then(jitsiLocalTrack => {
+                        closureLocalTrack = jitsiLocalTrack;
+                    })
+                    .then(
+                        () => new TrackVADEmitter(procNodeSampleRate, publishVAD, closureRnnoiseProc, closureLocalTrack)
+                    )
+            )
+            .catch(error => {
+                logger.error(`Failed to create TrackVADEmitter for ${micDeviceId} with error: ${error}`);
+
+                if (closureRnnoiseProc) {
+                    closureRnnoiseProc.destroy();
+                }
+
+                if (closureLocalTrack) {
+                    closureLocalTrack.stopStream();
+                }
+
+                throw error;
+            });
+    }
+
+    /**
+     * Sets up the audio graph in the AudioContext.
+     *
+     * @returns {Promise<void>}
+     */
+    _initializeAudioContext() {
+        this._audioSource = this._audioContext.createMediaStreamSource(this._jitsiLocalTrack.stream);
+
+        // TODO AudioProcessingNode is deprecated check and replace with alternative.
+        // We don't need stereo for determining the VAD score so we create a single chanel processing node.
+        this._audioProcessingNode = this._audioContext.createScriptProcessor(this._procNodeSampleRate, 1, 1);
+        this._audioProcessingNode.onaudioprocess = this._onAudioProcess;
+    }
+
+    /**
+     * ScriptProcessorNode callback, the input parameters contains the PCM audio that is then sent to rnnoise.
+     * Rnnoise only accepts PCM samples of 480 bytes whereas the webaudio processor node can't sample at a multiple
+     * of 480 thus after each _onAudioProcess callback there will remain and PCM buffer residue equal
+     * to _procNodeSampleRate / 480 which will be added to the next sample buffer and so on.
+     *
+     * @param {AudioProcessingEvent} audioEvent - Audio event.
+     * @returns {void}
+     */
+    _onAudioProcess(audioEvent: Object) {
+        // Prepend the residue PCM buffer from the previous process callback.
+        const inData = audioEvent.inputBuffer.getChannelData(0);
+        const completeInData = [ ...this._bufferResidue, ...inData ];
+        const sampleTimestamp = Date.now();
+
+        let i = 0;
+
+        for (; i + RNNOISE_SAMPLE_LENGTH < completeInData.length; i += RNNOISE_SAMPLE_LENGTH) {
+            const pcmSample = completeInData.slice(i, i + RNNOISE_SAMPLE_LENGTH);
+            const vadScore = this._rnnoiseProcessor.calculateAudioFrameVAD(pcmSample);
+
+            this._publishVAD({ timestamp: sampleTimestamp,
```

Comment:
Well I intentionally did that because the purpose of the timestamp was to show when the actual signal was recorded, and was to be used for potential time sync functions between different streams. Making it relative to process time defeats the purpose of what I had envisioned.